### PR TITLE
gh-119182: Add PyUnicodeWriter_WriteUCS4() function

### PR DESCRIPTION
**Pull request by vstinner**
* Issue: gh-119182

----
📚 Documentation preview 📚: https://cpython-previews--120849.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
cc @serhiy-storchaka @encukou

File: Doc/c-api/unicode.rst [hunk 1]
```diff
@@ -1563,6 +1563,16 @@ object.
    On success, return ``0``.
    On error, set an exception, leave the writer unchanged, and return ``-1``.
 
+.. c:function:: int PyUnicodeWriter_WriteUCS4(PyUnicodeWriter *writer, Py_UCS4 *str, Py_ssize_t size)
+
+   Writer the UCS4 string *str* into *writer*.
+
+   *size* is a number of UCS4 characters. If *size* is equal to ``-1``, get the
+   string length (search the NUL character).
```
Comment (serhiy-storchaka) [inline Doc/c-api/unicode.rst]: Is there a use case for null-terminated UCS4 string? `_PyUnicode_FromUCS4()` does not support this, as well as `PyUnicode_FromKindAndData()`. If you use the UCS4 buffer to build a string, you know how much characters were written into it.

Note also that `_PyUnicode_FromUCS4()` is not public. The user is supposed to use more general `PyUnicode_FromKindAndData()`.

Comment (vstinner) [reply Doc/c-api/unicode.rst]: > Is there a use case for null-terminated UCS4 string?

I don't know. I removed this code: size must be positive.

File: Lib/test/test_capi/test_unicode.py [hunk 2]
```diff
@@ -1784,8 +1784,24 @@ def test_widechar(self):
...
+        writer.write_char("-")
+        writer.write_ucs4("max=\U0010ffff", -1)
         writer.write_char('.')
-        self.assertEqual(writer.finish(), "latin1=\xE9-euro=\u20AC.")
+        self.assertEqual(writer.finish(),
+                         "latin1=\xE9-euro=\u20AC-max=\U0010ffff.")
+
+    def test_ucs4(self):
+        writer = self.create_writer(0)
+        writer.write_ucs4("ascii", -1)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_unicode.py]: All tests use negative size. I suppose they all will fail.

Please add also tests with embedded null characters, with lone a surrogate pair, with  surrogates, with non null-terminated string (`size` lesser that the string size) and with invalid size.

Comment (vstinner) [reply Lib/test/test_capi/test_unicode.py]: Ok, I added more tests.

> All tests use negative size. I suppose they all will fail.

Oops, fixed.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

**Comment by vstinner** [comment]
Merged, thanks for the review @serhiy-storchaka.

**Pull request merged**